tests/tcp.rs

```diff
@@ -79,3 +79,52 @@ fn accept2() {
     mine.unwrap();
     t.join().unwrap();
 }
+
+#[cfg(unix)]
+mod unix {
+    use tokio::net::TcpStream;
+    use tokio::prelude::*;
+
+    use env_logger;
+    use futures::future;
+    use mio::unix::UnixReady;
+
+    use std::{net, thread};
+    use std::time::Duration;
+
+    #[test]
+    fn poll_hup() {
+        drop(env_logger::init());
+
+        let srv = t!(net::TcpListener::bind("127.0.0.1:0"));
+        let addr = t!(srv.local_addr());
+        let t = thread::spawn(move || {
+            let mut client = t!(srv.accept()).0;
+            client.write(b"hello world").unwrap();
+            thread::sleep(Duration::from_millis(200));
+        });
+
+        let mut stream = t!(TcpStream::connect(&addr).wait());
+
+        // Poll for HUP before reading.
+        future::poll_fn(|| {
+            stream.poll_read_ready(UnixReady::hup().into())
+        }).wait().unwrap();
```

Comment:
i'm not entirely clear on what this tests validates:

as i read this, i expect that this blocks until `stream.poll_ready_ready(HUP)` returns ready?  Why should it return ready here?

If I'm misreading, what does it do?  Should we test that polling with a hup max actually notifies on a hup?

Reply:
The other side of the stream writes some data and closes the socket. This is testing that one can receive HUP notifications **before** reading the data from the socket.